File: pylib/slfhst/healthcheck.py
"""Stage2: final sanity check + MOTD banner summarizing the live stack."""
from __future__ import annotations

from pathlib import Path

from . import common, monitor
from .common import log

MOTD_FILE = Path("/etc/motd.d/slfhst.motd")


def write_motd(cfg: dict) -> None:
    domain = cfg.get("domain", "<domain>")
    lines = [
        "",
        "slfhst is up. Services:",
        f"  mail    https://mail.{domain}",
        f"  vault   https://vault.{domain}",
        f"  photos  https://photos.{domain}",
        f"  auth    https://auth.{domain}",
        f"  locker  https://locker.{domain}",
        "",
        "TOTP enrollment (secret/QR/recovery codes) was shown ONLY on the",
        "physical console during setup -- it is not stored anywhere else.",
        "",
    ]
    MOTD_FILE.parent.mkdir(parents=True, exist_ok=True)
    MOTD_FILE.write_text("\n".join(lines))


def main() -> None:
    common.require_done_or_exit("stage1")
    cfg = common.load_config()

    statuses = monitor.service_status()
    down = [name for name, state in statuses.items() if state != "active"]
    if down:
        log.warning("services not active yet (may still be starting): %s", ", ".join(down))
    else:
        log.info("all app services active")

    write_motd(cfg)


if __name__ == "__main__":
    main()

File: internal/healthcheck/healthcheck.go
// Package healthcheck is stage2's final sanity check + MOTD banner.
// Direct port of healthcheck.py.
package healthcheck

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/nnnithinnn/slfhst/internal/config"
	"github.com/nnnithinnn/slfhst/internal/monitor"
)

const motdFile = "/etc/motd.d/slfhst.motd"

// writeMOTD writes the service-URL banner shown at SSH login. Direct
// port of healthcheck.py's write_motd() -- note api.<domain> (Ente's
// backend) is deliberately not listed, matching the Python original.
func writeMOTD(cfg map[string]any) error {
	domain, _ := cfg["domain"].(string)
	lines := fmt.Sprintf(`slfhst -- self-hosted mail + photos/auth/locker + vault

  Mail:   https://mail.%[1]s
  Vault:  https://vault.%[1]s
  Photos: https://photos.%[1]s
  Auth:   https://auth.%[1]s
  Locker: https://locker.%[1]s

TOTP enrollment was shown once on the console during setup and isn't
stored anywhere else -- if you lost it, re-enroll via google-authenticator.
`, domain)

	if err := os.MkdirAll(filepath.Dir(motdFile), 0o755); err != nil {
		return fmt.Errorf("healthcheck: mkdir %s: %w", filepath.Dir(motdFile), err)
	}
	if err := os.WriteFile(motdFile, []byte(lines), 0o644); err != nil {
		return fmt.Errorf("healthcheck: write %s: %w", motdFile, err)
	}
	return nil
}

// Run is stage2's final step: log which app services aren't active yet
// (a warning, not a failure -- containers can still be starting), then
// write the MOTD. Direct port of healthcheck.py's main().
func Run(store config.Store) error {
	if err := store.RequireDone("stage1"); err != nil {
		return err
	}
	cfg, err := store.Load()
	if err != nil {
		return err
	}

	statuses := monitor.ServiceStatus()
	var notActive []string
	for name, status := range statuses {
		if status != "active" {
			notActive = append(notActive, fmt.Sprintf("%s=%s", name, status))
		}
	}
	if len(notActive) > 0 {
		fmt.Println("warning: healthcheck: not all app services are active yet:", notActive)
	} else {
		fmt.Println("healthcheck: all app services active")
	}

	return writeMOTD(cfg)
}
